File: src/RandomAgent.cpp
#include "RandomAgent.hpp"
#include "Agent.hpp"
#include "AES.hpp"

RandomAgent::RandomAgent(Agent *agent, size_t bufferSize)
: agent(agent), bufferSize(bufferSize)
{}

RandomAgent::~RandomAgent()
{}

void RandomAgent::FillBuffer()
{
	while (buffer.size() < bufferSize) {
		int64_t bid = rand() % agent->GetBlockCount();

		block data = agent->Access(Agent::READ, bid, {});

		ChangedMap changed = agent->LoadChangedMap();
		changed[bid] = 0;
		agent->SaveChangedMap(changed);

		buffer[bid] = data;
	}
}

int64_t RandomAgent::GetBlock()
{
	ChangedMap changed = agent->LoadChangedMap();
	
	// Assumes atleast 1 block hasn't been changed
	for (;;) {
		// Pick a random element
		auto iter = buffer.begin();
		std::advance(iter, rand() % buffer.size());
		
		size_t bid = iter->first;

		if (changed[bid] == 1) {
			// The data in the buffer is old
			buffer.erase(iter);

			// We've acknowledged the data has changed
			changed[bid] = 0;
		} else {
			// Save any changes we may have made
			agent->SaveChangedMap(changed);
		
			return bid;
		}
	}
}


void RandomAgent::Run()
{
	FillBuffer();

	int64_t bid = GetBlock();
	agent->Duplicate(bid, buffer[bid]);
}

File: lua/random_agent.lua
require "agent"

buffer = {}
buffer_length = 5

function fill_buffer(blockmap)
	while #buffer < buffer_length do
		local bid = math.random(10)
	
		changed[bid] = false

		local elem = {}
		elem.bid = bid
		elem.data = access("READ", bid, blockmap)
		--print(elem.data)
		table.insert(buffer, elem)
	end
end

function random_agent(blockmap)
	fill_buffer(blockmap)
	
	-- Pick a random element
	local index = random(buffer)
	local elem = buffer[index]
	
	-- Randomise the block
	elem.data = encrypt(decrypt(elem.data))
	
	-- Write the data
	local posmap = build_posmap(blockmap)
	local new_pos = takePosition(posmap, blockmap)
	server_write(new_pos, elem.data)

	if changed[elem.bid] then
		-- Remove element from buffer
		buffer[index] = buffer[#buffer]
		buffer[#buffer] = nil
		
		changed[elem.bid] = false
	else
		-- It's adding to the set for that block
		table.insert(blockmap[elem.bid], new_pos)
	end
end
